runtime/src/main/java/org/corfudb/runtime/collections/StreamSubscription.java

```diff
@@ -127,6 +129,11 @@ boolean enqueueStreamEntry(ILogData logData, long maxBlockTime) throws Interrupt
                     .stream()
                     .map(entry -> CorfuStreamEntry.fromSMREntry(entry, epoch))
                     .collect(Collectors.toList());
+
+            // Deduplicate entries per stream Id
+            Set<Message> observedKeys = new HashSet<>();
+            entryList.removeIf(e-> !observedKeys.add(e.getKey()));
```

Comment:
Very neat!

Reply:
I think this is wrong, since it will keep the older observed key not the newest one. You need to iterate the `entryList` backwards to keep the latest keys.

Reply:
For example if  `|E1 | E2 | E3` all have the same key, then the algorithm will filter E2 and E3, but it should have kept E3 and filtered out the rest.

Reply:
Correct, I had already changed it in my second push, so closing as I addressed this and added the consideration in the test.